### ISSUE TITLE
Injecting types defined in parent folders (and their children)

### ISSUE THREAD

**Issue by ZoolWay** [aurelia/dependency-injection#104]
I got a problem when the definition of a type which I want to inject is in a parent folder (or one of the parent's sub folders).

I am importing my service like
`import { ConfigurationService } from '../core/configuration-service'`

But neither `@inject` nor `@autoinject` will work. I always get a `badkeyerror`. I tracked it down to a call in aurelia-dependency-injection to `_aureliaMetadata.metadata.getOwn("design:paramtypes", target) || _emptyParameters` which returns an array with an undefined entry for this case.

When using classes from the current folder or a subfolder of the current folder it works, e.g.:
`import { AnotherService } from './another-service'` 
or
`import { PartyService } from './party/party-service'`

I am using RC1 with TypeScript 1.8.10 targetting ES6, working from skeleton with systemjs/jspm.

Am I refering my types wrongly, is this a bug with metadata or with DI?

**Comment by ZoolWay** [comment]
I have tried to adopt all import statements to paths relative to the src folder, but that does not help either.

**Comment by EisenbergEffect** [comment]
This isn't an Aurelia problem. If you are gettng that error, it's because the import itself is incorrect.

**Comment by ZoolWay** [comment]
Let's say I have these components in my aurelia app:

app.ts
core/configuration-service.ts
users/user-service.ts

How can I import `ConfigurationService` inside `UserService`?
For the TypeScript compiler both of these are fine:
```
import { ConfigurationService } from 'core/configuration-service.ts';
import { ConfigurationService } from '../core/configuration-service.ts';
```

Also if I do such an import, I can create an instance with the `new` operator. It is just DI failing to provide it and only if the class is not in the same folder (or a subfolder of it).

So if I can do a `new`, how is the import wrong?

**Comment by EisenbergEffect** [comment]
Show me the code for the class where you want the injection to happen.

**Comment by ZoolWay** [comment]
Okay, this is `brick-service.ts` in subfolder `bricks`.

This works, not using DI:
```
import { autoinject, inject, LogManager } from 'aurelia-framework';
import { HttpClient } from 'aurelia-fetch-client';

import { BrickDef } from 'bricks/brick-def';
import { BrickGroup } from 'bricks/brick-group';
import { GateService } from 'core/gate-service';
import { ConfigurationService } from 'core/configuration-service';

export class BrickService {
    private log = LogManager.getLogger("loggerName");
    private client: HttpClient;
    private configurationService: ConfigurationService;
    private gateService: GateService;
    constructor() {
        this.configurationService = new ConfigurationService(); // TODO replace with DI
        this.gateService = new GateService(this.configurationService); // TODO replace with DI
        this.log.debug("Got objects:", this.configurationService, this.gateService);
        debugger;
    }
    public getAvailableBrickDefinitions(): BrickDef[] {
        return new BrickDef[0];
    }
    public getDefinitionById(id: string): BrickDef {
        return null;
    }
    public async initRemoteBricks(): Promise<boolean> {
        return true;
    }
}
```

Using DI for `GateService` and `ConfigurationService`, both in the sibling directory `core`, I get the badKeyError:

```
import { autoinject, inject, LogManager } from 'aurelia-framework';
import { HttpClient } from 'aurelia-fetch-client';

import { BrickDef } from 'bricks/brick-def';
import { BrickGroup } from 'bricks/brick-group';
import { GateService } from 'core/gate-service';
import { ConfigurationService } from 'core/configuration-service';

@autoinject
export class BrickService {

    private log = LogManager.getLogger("loggerName");
    private client: HttpClient;
    private configurationService: ConfigurationService;
    private gateService: GateService;

    constructor(configurationService: ConfigurationService, gateService: GateService) {
        this.configurationService = configurationService;
        this.gateService = gateService;
        this.log.debug("Got objects:", this.configurationService, this.gateService);
        debugger;
    }
    public getAvailableBrickDefinitions(): BrickDef[] {
        return new BrickDef[0];
    }
    public getDefinitionById(id: string): BrickDef {
        return null;
    }
    public async initRemoteBricks(): Promise<boolean> {
        return true;
    }
}
```

I have also tried to use `@inject(ConfigurationService, GateService)` instead of `@autoinject`.

It always throw an `Error(badKeyError)` in `Container.get`.

This does not happen when all classes/modules are in the same folder!

**Comment by EisenbergEffect** [comment]
Do you have circular references?

**Comment by EisenbergEffect** [comment]
The bad key error happens because the thing you are trying to inject is undefined. That means that the import isn't working correctly. (Nothing to do with Aurelia at that point.)

**Comment by ZoolWay** [comment]
I can exchange that dependencies for a newly created `TestService` from the `core` folder which is empty and so has no dependencies. The same problem.

Yes, the thing I am trying to import is undefined but I have no idea at this point how to debug that. Importing from other modules in my own app should in fact be very simple.

Maybe I messed with the module loading or something, I will try to further investigate.

**Comment by ZoolWay** [comment]
But which would be the correct syntax for the import in your opinion?
```
a) import { ConfigurationService } from 'core/configuration-service';
b) import { ConfigurationService } from '../core/configuration-service';
```

**Comment by EisenbergEffect** [comment]
It depends on how your module loader is configured. 
Can you zip up a simple version of your project that produces the issue so I can take a look?

**Comment by ZoolWay** [comment]
Thanks for your support Rob!

While I was stripping all unnecessary things from my project to provide you a small sample - of course - I did also strip what was causing the issue.

Sorry that I bothered, sure thing I messed up the module loading.

This is how:

As you can see from a sample above I got some classes BrickGroup, BrickDef etc. which all have their very own file because I like the one class one file approach.
But that approach can mean lots of lines of import at the beginning of a file. So I basically created a `brick-module.ts` which does just this:
```
export * from 'bricks/brick-def';
export * from 'bricks/brick-group';
...
```

So I can import like this: `import { BrickDef, BrickGroup } from 'bricks/brick-module';`

And of course I was mixing that with importing that classes directly from their source files. Guess I will drop that and go with more `import` statements.

Shame on me :-(

**Comment by EisenbergEffect** [comment]
Some of the transpilers have various issues with export * I think.

**Comment by ZoolWay** [comment]
Your answer made me courious.

I retried changing my `brick-module.ts` not to use the star `*` but the class names. But that did not the trick.

Further I found out that not all places would have to omit brick-module.ts and go for the *real* modules to have it back to work. But I prefer to have it all consistent.

I guess even when imported from the same module, the `BrickService` from `brick-service.ts` and the one from `brick-module.ts` are not considered to be the same in all cases.
